src/storage/mutate/AddVerticesProcessor.cpp

```diff
@@ -180,134 +170,138 @@ void AddVerticesProcessor::doProcessWithIndex(const cpp2::AddVerticesRequest& re
-                  }
-                }
+        // step 2, Insert new vertex index
+        if (newReader != nullptr) {
+          auto newIndexKeys = indexKeys(partId, vId.str(), newReader.get(), index, nullptr);
+          if (!newIndexKeys.empty()) {
+            // check if index has ttl field, write it to index value if exists
+            auto field = CommonUtils::ttlValue(schema.get(), newReader.get());
+            auto indexVal = field.ok() ? IndexKeyUtils::indexVal(std::move(field).value()) : "";
+            auto indexState = env_->getIndexState(spaceId_, partId);
+            if (env_->checkRebuilding(indexState)) {
+              for (auto& idxKey : newIndexKeys) {
+                auto opKey = OperationKeyUtils::modifyOperationKey(partId, idxKey);
+                ret.writeSet.emplace_back(std::string(opKey));
+                batchHolder->put(std::move(opKey), std::string(indexVal));
+              }
+            } else if (env_->checkIndexLocked(indexState)) {
+              LOG(ERROR) << "The index has been locked: " << index->get_index_name();
+              ret.code = nebula::cpp2::ErrorCode::E_DATA_CONFLICT_ERROR;
+              return ret;
+            } else {
+              for (auto& idxKey : newIndexKeys) {
+                ret.writeSet.emplace_back(std::string(idxKey));
+                batchHolder->put(std::move(idxKey), std::string(indexVal));
               }
             }
           }
-        }  // for index data
-        if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
-          break;
         }
-        /*
-         * step 3 , Insert new vertex data
-         */
-        batchHolder->put(std::move(key), std::move(retEnc.value()));
-        stats::StatsManager::addValue(kNumVerticesInserted);
       }
-      if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
-        break;
-      }
-    }
-    if (code != nebula::cpp2::ErrorCode::SUCCEEDED) {
-      env_->verticesML_->unlockBatch(dummyLock);
-      handleAsync(spaceId_, partId, code);
-      continue;
     }
-    auto batch = encodeBatchValue(batchHolder->getBatch());
-    DCHECK(!batch.empty());
-    nebula::MemoryLockGuard<VMLI> lg(env_->verticesML_.get(), std::move(dummyLock), false, false);
-    env_->kvstore_->asyncAppendBatch(spaceId_,
-                                     partId,
-                                     std::move(batch),
-                                     [l = std::move(lg), icw = std::move(wrapper), partId, this](
-                                         nebula::cpp2::ErrorCode retCode) {
-                                       UNUSED(l);
-                                       UNUSED(icw);
-                                       handleAsync(spaceId_, partId, retCode);
-                                     });
+    // step 3, Insert new vertex data
+    ret.writeSet.emplace_back(key);
+    batchHolder->put(std::move(key), std::move(value));
   }
-}  // namespace storage
+  ret.batch = encodeBatchValue(batchHolder->getBatch());
```

Comment:
I'm considering whether we could move all log-encoding into raft later, because we need to first encode here, and decode in raft again, it also introduce some extra string copy.